Enum with struct variants can only have 12 fields

### Bug Description

I was refactoring some code from 

```rust
#[pyclass]
enum TupleEnum {
   Variant1(InnerType1),
   Variant2(InnerType2),
}

#[pyclass]
struct InnerType1 {
    field_1: u32,
    field_2: u32,
    field_3: String,
    field_4: f32,
}

...

```

to this

```rust
#[pyclass]
enum StructEnum {
    Variant1 {
        field_1: u32,
        field_2: u32,
        field_3: String,
        field_4: f32,
    },
    Variant2 {
        etc...
    }
}
```

Effectively getting rid of the wrapped types and representing them in the enum's variants directly.

However, it looks like there is some kind of limit on the number of fields that you can have in an enum member. The example in the "Steps to reproduce" produces this error:

```
error[E0277]: `(&str, &str, &str, &str, &str, &str, &str, &str, &str, &str, &str, &str, &str, &str, &str)` cannot be converted to a Python object
   --> pyanif/src/blocks/beam.rs:373:1
    |
373 | #[pyclass]
    | ^^^^^^^^^^ the trait `pyo3::IntoPyObject<'_>` is not implemented for `(&str, &str, &str, &str, &str, &str, &str, &str, &str, &str, &str, &str, &str, &str, &str)`
    |
    = note: `IntoPyObject` is automatically implemented by the `#[pyclass]` macro
    = note: if you do not wish to have a corresponding Python type, implement it manually
    = note: if you do not own `(&str, &str, &str, &str, &str, &str, &str, &str, &str, &str, &str, &str, &str, &str, &str)` you can perform a manual conversion to one of the types in `pyo3::types::*`
    = help: the following other types implement trait `pyo3::IntoPyObject<'py>`:
              &(T0, T1)
              &(T0, T1, T2)
              &(T0, T1, T2, T3)
              &(T0, T1, T2, T3, T4)
              &(T0, T1, T2, T3, T4, T5)
              &(T0, T1, T2, T3, T4, T5, T6)
              &(T0, T1, T2, T3, T4, T5, T6, T7)
              &(T0, T1, T2, T3, T4, T5, T6, T7, T8)
            and 17 others
note: required by a bound in `pyo3::IntoPyObjectExt::into_py_any`
   --> /home/majp/.cargo/registry/src/index.crates.io-6f17d22bba15001f/pyo3-0.23.3/src/conversion.rs:367:33
    |
367 | pub trait IntoPyObjectExt<'py>: IntoPyObject<'py> + into_pyobject_ext::Sealed {
    |                                 ^^^^^^^^^^^^^^^^^ required by this bound in `IntoPyObjectExt::into_py_any`
...
380 |     fn into_py_any(self, py: Python<'py>) -> PyResult<Py<PyAny>> {
    |        ----------- required by a bound in this associated function
    = note: this error originates in the attribute macro `pyclass` (in Nightly builds, run with -Z macro-backtrace for more info)

```


### Steps to Reproduce

```rust
// my_module.rs

use pyo3::prelude::*

enum StructEnum {
    Variant1 {
        field_1: u32,
        field_2: f32,
        field_3: String,
        field_4: f32,
        field_5: f32,
        field_6: u32,
        field_7: u32,
        field_8: u32,
        field_9: u32,
        field_10: u32,
        field_11: u32,
        field_12: u32,
        field_13: u32,
        field_14: u32,
        field_15: u32,
    }
}
```

won't compile. However, with 12 fields, there is no error:

```rust
#[pyclass]
enum StructEnum {
    Variant1 {
        field_1: u32,
        field_2: f32,
        field_3: String,
        field_4: f32,
        field_5: f32,
        field_6: u32,
        field_7: u32,
        field_8: u32,
        field_9: u32,
        field_10: u32,
        field_11: u32,
        field_12: u32,
        // field_13: u32,
        // field_14: u32,
        // field_15: u32,
    }
}
```

### Backtrace

```shell
error[E0277]: `(&str, &str, &str, &str, &str, &str, &str, &str, &str, &str, &str, &str, &str, &str, &str)` cannot be converted to a Python object
   --> pyanif/src/blocks/beam.rs:373:1
    |
373 | #[pyclass]
    | ^^^^^^^^^^ the trait `pyo3::IntoPyObject<'_>` is not implemented for `(&str, &str, &str, &str, &str, &str, &str, &str, &str, &str, &str, &str, &str, &str, &str)`
    |
    = note: `IntoPyObject` is automatically implemented by the `#[pyclass]` macro
    = note: if you do not wish to have a corresponding Python type, implement it manually
    = note: if you do not own `(&str, &str, &str, &str, &str, &str, &str, &str, &str, &str, &str, &str, &str, &str, &str)` you can perform a manual conversion to one of the types in `pyo3::types::*`
    = help: the following other types implement trait `pyo3::IntoPyObject<'py>`:
              &(T0, T1)
              &(T0, T1, T2)
              &(T0, T1, T2, T3)
              &(T0, T1, T2, T3, T4)
              &(T0, T1, T2, T3, T4, T5)
              &(T0, T1, T2, T3, T4, T5, T6)
              &(T0, T1, T2, T3, T4, T5, T6, T7)
              &(T0, T1, T2, T3, T4, T5, T6, T7, T8)
            and 17 others
note: required by a bound in `pyo3::IntoPyObjectExt::into_py_any`
   --> /home/majp/.cargo/registry/src/index.crates.io-6f17d22bba15001f/pyo3-0.23.3/src/conversion.rs:367:33
    |
367 | pub trait IntoPyObjectExt<'py>: IntoPyObject<'py> + into_pyobject_ext::Sealed {
    |                                 ^^^^^^^^^^^^^^^^^ required by this bound in `IntoPyObjectExt::into_py_any`
...
380 |     fn into_py_any(self, py: Python<'py>) -> PyResult<Py<PyAny>> {
    |        ----------- required by a bound in this associated function
    = note: this error originates in the attribute macro `pyclass` (in Nightly builds, run with -Z macro-backtrace for more info)


```


### Your operating system and version

Ubuntu 24.04

### Your Python version (`python --version`)

Python 3.12.3

### Your Rust version (`rustc --version`)

rustc 1.83.0 (90b35a623 2024-11-26)

### Your PyO3 version

0.23.3

### How did you install python? Did you use a virtualenv?

No venv, just the `python3` that comes with Ubuntu 24.04.

### Additional Info

Thanks for this awesome project. I imagine creating this is an insane task. So lots of appreciation from here :)

--- a/newsfragments/4832.fixed.md
+++ b/newsfragments/4832.fixed.md
@@ -1,0 +1,1 @@
+`#[pyclass]` complex enums support more than 12 variant fields.

--- a/pyo3-macros-backend/src/pyclass.rs
+++ b/pyo3-macros-backend/src/pyclass.rs
@@ -19,8 +19,9 @@ use crate::method::{FnArg, FnSpec, PyArg, RegularArg};
 use crate::pyfunction::ConstructorAttribute;
 use crate::pyimpl::{gen_py_const, get_cfg_attributes, PyClassMethodsType};
 use crate::pymethod::{
-    impl_py_getter_def, impl_py_setter_def, MethodAndMethodDef, MethodAndSlotDef, PropertyType,
-    SlotDef, __GETITEM__, __HASH__, __INT__, __LEN__, __REPR__, __RICHCMP__, __STR__,
+    impl_py_class_attribute, impl_py_getter_def, impl_py_setter_def, MethodAndMethodDef,
+    MethodAndSlotDef, PropertyType, SlotDef, __GETITEM__, __HASH__, __INT__, __LEN__, __REPR__,
+    __RICHCMP__, __STR__,
 };
 use crate::pyversions::is_abi3_before;
 use crate::utils::{self, apply_renaming_rule, Ctx, LitCStr, PythonDoc};
@@ -1185,34 +1186,30 @@ fn impl_complex_enum_variant_cls(
 }
 
 fn impl_complex_enum_variant_match_args(
-    ctx: &Ctx,
+    ctx @ Ctx { pyo3_path, .. }: &Ctx,
     variant_cls_type: &syn::Type,
     field_names: &mut Vec<Ident>,
-) -> (MethodAndMethodDef, syn::ImplItemConst) {
+) -> syn::Result<(MethodAndMethodDef, syn::ImplItemFn)> {
     let ident = format_ident!("__match_args__");
-    let match_args_const_impl: syn::ImplItemConst = {
-        let args_tp = field_names.iter().map(|_| {
-            quote! { &'static str }
-        });
+    let mut match_args_impl: syn::ImplItemFn = {
         parse_quote! {
-            #[allow(non_upper_case_globals)]
-            const #ident: ( #(#args_tp,)* ) = (
-                #(stringify!(#field_names),)*
-            );
+            #[classattr]
+            fn #ident(py: #pyo3_path::Python<'_>) -> #pyo3_path::PyResult<#pyo3_path::Bound<'_, #pyo3_path::types::PyTuple>> {
+                #pyo3_path::types::PyTuple::new::<&str, _>(py, [
+                    #(stringify!(#field_names),)*
+                ])
+            }
         }
     };
 
-    let spec = ConstSpec {
-        rust_ident: ident,
-        attributes: ConstAttributes {
-            is_class_attr: true,
-            name: None,
-        },
-    };
+    let spec = FnSpec::parse(
+        &mut match_args_impl.sig,
+        &mut match_args_impl.attrs,
+        Default::default(),
+    )?;
+    let variant_match_args = impl_py_class_attribute(variant_cls_type, &spec, ctx)?;
 
-    let variant_match_args = gen_py_const(variant_cls_type, &spec, ctx);
-
-    (variant_match_args, match_args_const_impl)
+    Ok((variant_match_args, match_args_impl))
 }
 
 fn impl_complex_enum_struct_variant_cls(
@@ -1260,7 +1257,7 @@ fn impl_complex_enum_struct_variant_cls(
     }
 
     let (variant_match_args, match_args_const_impl) =
-        impl_complex_enum_variant_match_args(ctx, &variant_cls_type, &mut field_names);
+        impl_complex_enum_variant_match_args(ctx, &variant_cls_type, &mut field_names)?;
 
     field_getters.push(variant_match_args);
 
@@ -1268,6 +1265,7 @@ fn impl_complex_enum_struct_variant_cls(
         #[doc(hidden)]
         #[allow(non_snake_case)]
         impl #variant_cls {
+            #[allow(clippy::too_many_arguments)]
             fn __pymethod_constructor__(py: #pyo3_path::Python<'_>, #(#fields_with_types,)*) -> #pyo3_path::PyClassInitializer<#variant_cls> {
                 let base_value = #enum_name::#variant_ident { #(#field_names,)* };
                 <#pyo3_path::PyClassInitializer<#enum_name> as ::std::convert::From<#enum_name>>::from(base_value).add_subclass(#variant_cls)
@@ -1434,7 +1432,7 @@ fn impl_complex_enum_tuple_variant_cls(
     slots.push(variant_getitem);
 
     let (variant_match_args, match_args_method_impl) =
-        impl_complex_enum_variant_match_args(ctx, &variant_cls_type, &mut field_names);
+        impl_complex_enum_variant_match_args(ctx, &variant_cls_type, &mut field_names)?;
 
     field_getters.push(variant_match_args);
 
@@ -1442,6 +1440,7 @@ fn impl_complex_enum_tuple_variant_cls(
         #[doc(hidden)]
         #[allow(non_snake_case)]
         impl #variant_cls {
+            #[allow(clippy::too_many_arguments)]
             fn __pymethod_constructor__(py: #pyo3_path::Python<'_>, #(#field_names : #field_types,)*) -> #pyo3_path::PyClassInitializer<#variant_cls> {
                 let base_value = #enum_name::#variant_ident ( #(#field_names,)* );
                 <#pyo3_path::PyClassInitializer<#enum_name> as ::std::convert::From<#enum_name>>::from(base_value).add_subclass(#variant_cls)

--- a/pyo3-macros-backend/src/pymethod.rs
+++ b/pyo3-macros-backend/src/pymethod.rs
@@ -526,7 +526,7 @@ fn impl_clear_slot(cls: &syn::Type, spec: &FnSpec<'_>, ctx: &Ctx) -> syn::Result
     })
 }
 
-fn impl_py_class_attribute(
+pub(crate) fn impl_py_class_attribute(
     cls: &syn::Type,
     spec: &FnSpec<'_>,
     ctx: &Ctx,

--- a/src/tests/hygiene/pyclass.rs
+++ b/src/tests/hygiene/pyclass.rs
@@ -92,6 +92,44 @@ pub enum TupleEnumEqOrd {
     Variant2(u32),
 }
 
+#[crate::pyclass(crate = "crate")]
+pub enum ComplexEnumManyVariantFields {
+    ManyStructFields {
+        field_1: u16,
+        field_2: u32,
+        field_3: u32,
+        field_4: i32,
+        field_5: u32,
+        field_6: u32,
+        field_7: u8,
+        field_8: u32,
+        field_9: i32,
+        field_10: u32,
+        field_11: u32,
+        field_12: u32,
+        field_13: u32,
+        field_14: i16,
+        field_15: u32,
+    },
+    ManyTupleFields(
+        u16,
+        u32,
+        u32,
+        i32,
+        u32,
+        u32,
+        u8,
+        u32,
+        i32,
+        u32,
+        u32,
+        u32,
+        u32,
+        i16,
+        u32,
+    ),
+}
+
 #[crate::pyclass(str = "{x}, {y}, {z}")]
 #[pyo3(crate = "crate")]
 pub struct PointFmt {

